guppylang/tys/ty.py

```diff
@@ -667,10 +667,11 @@ def unify(s: Type | Const, t: Type | Const, subst: "Subst | None") -> "Subst | N
         case NoneType(), NoneType():
             return subst
         case FunctionType() as s, FunctionType() as t if s.params == t.params:
-            if len(s.inputs) != len(t.inputs) or any(
-                a.flags != b.flags for a, b in zip(s.inputs, t.inputs, strict=True)
-            ):
+            if len(s.inputs) != len(t.inputs):
                 return None
+            for a, b in zip(s.inputs, t.inputs, strict=True):
+                if a.ty.linear and b.ty.linear and a.flags != b.flags:
```

Comment:
I think we're going to end up with flags for non-linear types, so would it be better to just implicitly add `@owned` to non-linear types and hide it when printing (for now)?

Reply:
This was introduced in #531